src/module_3/push_exploration.md

```diff
@@ -0,0 +1,5364 @@
+    
+               0       0.99      0.97      0.98    213182
+               1       0.14      0.35      0.20      3213
+    
+        accuracy                           0.96    216395
+       macro avg       0.56      0.66      0.59    216395
+    weighted avg       0.98      0.96      0.97    216395
+    
+    Confusion Matrix:
+    [[206210   6972]
+     [  2083   1130]]
+
+
+Pretty much same results as before. Lasso took 2h to run & didn't reduced any additonal coeff to 0, obtaining very similar results. This might mean that there is not any irrelevant features among the selected.
+
+Let's try then to play with alpha parameter for stronger regularizations. If alpha increses, then the regularization is stronger. This parameter is adjusted by the C parameter in scikit, which is the inverse of alpha (i.e. higher alpha == lower C)
+
+
+```python
+X = final_data.drop(columns='outcome')
+y = final_data['outcome']
+
+scaler = StandardScaler()
+X = scaler.fit_transform(X)
+
+# Split again
+X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.2, random_state=33)
+X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=33)
+
+model1 = LogisticRegression(C=0.5, class_weight='balanced')
+model2 = LogisticRegression(C=0.1, class_weight='balanced')
+model3 = LogisticRegression(C=0.01, class_weight='balanced')
+
+model1.fit(X_train, y_train)
+model2.fit(X_train, y_train)
+model3.fit(X_train, y_train)
+```
+
+
+```python
+# Models with different values of C
+models = [model1, model2, model3]
+colors = ['r', 'g', 'b']
+
+plt.figure(figsize=(10, 5))
+
+# For each model, calculate and plot ROC and Precision-Recall curves
+for i, model in enumerate(models):
+    # Calculate probabilities for the positive class
+    y_val_probs = model.predict_proba(X_val)[:, 1]
+
+    # ROC Curve
+    fpr_val, tpr_val, _ = roc_curve(y_val, y_val_probs)
+    roc_auc_val = auc(fpr_val, tpr_val)  # Calculate AUC
+
+    plt.subplot(1, 2, 1)
+    plt.plot(fpr_val, tpr_val, label=f'ROC (AUC = {roc_auc_val:.2f})', color=colors[i])
+
+    # Precision-Recall Curve
+    precision_val, recall_val, _ = precision_recall_curve(y_val, y_val_probs)
+
+    plt.subplot(1, 2, 2)
+    plt.plot(recall_val, precision_val, color=colors[i])
+
+plt.subplot(1, 2, 1)
+plt.plot([0, 1], [0, 1], color='navy', linestyle='--')
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('False Positive Rate')
+plt.ylabel('True Positive Rate')
+plt.title('ROC')
+plt.legend(loc="lower right")
+
+plt.subplot(1, 2, 2)
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('Recall')
+plt.ylabel('Precision')
+plt.title('Precision-Recall Curve')
+
+plt.tight_layout()
+plt.show()
+
+```
+
+
+    
+![png](push_exploration_files/push_exploration_71_0.png)
+    
+
+
+
+```python
+X = final_data.drop(columns='outcome')
+y = final_data['outcome']
+
+# Models with different values of C
+models = [model1, model2, model3]
+
+for i, model in enumerate(models):
```

Comment:
It is quite weird that all the models give you the exact same AUC. If you put the regularization parameter high enough so you have a very big regularization, at certain point the performance of the model should degrade, could you try it?